Ical.Net/Evaluation/RecurrencePatternEvaluator.cs

```diff
@@ -908,17 +908,17 @@ specified as a date with UTC time.
         {
             // If 'reference' has a timezone, 'until' MUST be UTC,
             // but in case of UTC rule violation we fall back to the 'reference' timezone
-            untilTzId = until.Kind == DateTimeKind.Utc
+            untilTzId = until.TzId == CalDateTime.UtcTzId
                 ? CalDateTime.UtcTzId
                 : reference.TzId;
         }
 
-        var untilCalDt = new CalDateTime(until, untilTzId, reference.HasTime);
+        var untilCalDt = new CalDateTime(until.Value, untilTzId, reference.HasTime);
 
         // If 'reference' is floating, then 'until' is floating, too
         return reference.TzId is null
-            ? untilCalDt.Value
-            // convert to the reference timezone
-            : untilCalDt.ToTimeZone(reference.TzId).Value;
+            ? untilCalDt
+            // convert to the reference timezone and convert the value to Floating
+            : untilCalDt.ToTimeZone(reference.TzId).Value.AsCalDateTime();
```

Comment:
We should have a more elegant way of converting between floating and non-floating. Can't we write `.ToTimeZone(null)`? I feel that we should actually have something like `CalDateTime.AsFloating()` and `CalDateTime.WithZoned(tzId)` (which is not a topic of this PR and which is in contrast to my comments in a previous discussion - I know)

Reply:
OMG, of course, this is implemented already: `ToTimeZone(reference.TzId).ToTimeZone(null)`, as simple as that 👍